src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogram.java

```diff
@@ -0,0 +1,284 @@
+    this.clockMillis = clockMillis;
+  }
+
+  public void update(int value) {
+    update((double) value);
+  }
+
+  public void update(long value) {
+    update((double) value);
+  }
+
+  public void update(double value) {
+    getCurrent().getDistribution().add(value);
+  }
+
+  /**
+   * Bulk-update this histogram with a set of centroids.
+   *
+   * @param means  the centroid values
+   * @param counts the centroid weights/sample counts
+   */
+  public void bulkUpdate(List<Double> means, List<Integer> counts) {
+    if (means != null && counts != null) {
+      int n = Math.min(means.size(), counts.size());
+      MinuteBin current = getCurrent();
+      for (int i = 0; i < n; ++i) {
+        current.getDistribution().add(means.get(i), counts.get(i));
+      }
+    }
+  }
+
+  /**
+   * @return returns the number of values in the distribution.
+   */
+  public long getCount() {
+    return perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        mapToLong(bin -> bin.getDistribution().size()).sum();
+  }
+
+  /**
+   * @return returns the maximum value in the distribution. Returns NaN if the distribution is empty.
+   */
+  public double getMax() {
+    return perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        mapToDouble(b -> b.getDistribution().getMax()).max().orElse(NaN);
+  }
+
+  /**
+   * @return returns the minimum value in the distribution. Returns NaN if the distribution is empty.
+   */
+  public double getMin() {
+    return perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        mapToDouble(b -> b.getDistribution().getMin()).min().orElse(NaN);
+  }
+
+  /**
+   * @return returns the mean of the values in the distribution. Returns NaN if the distribution is empty.
+   */
+  public double getMean() {
+    List<Centroid> centroids = new ArrayList<>();
+    perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        forEach(bin -> centroids.addAll(bin.getDistribution().centroids()));
+
+    return centroids.size() == 0 ?
+        NaN :
+        centroids.stream().mapToDouble(c -> (c.count() * c.mean()) / centroids.size()).sum();
+  }
+
+  /**
+   * @return returns the sum of the values in the distribution.
+   */
+  public double getSum() {
+    List<Centroid> centroids = new ArrayList<>();
+    perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        forEach(bin -> centroids.addAll(bin.getDistribution().centroids()));
+
+    return centroids.stream().mapToDouble(c -> c.count() * c.mean()).sum();
+  }
+
+  /**
+   * Aggregates all the bins prior to the current minute
+   * This is because threads might be updating the current minute bin while the bins() method is invoked
+   *
+   * @param clear if set to true, will clear all values in bins prior to the current minute
+   * @return returns aggregated collection of all the bins prior to the current minute
+   */
+  public List<MinuteBin> bins(boolean clear) {
```

Comment:
I don't think we should give the boolean option "clear" to the users of this SDK.
Any aggregated values prior to the current min should be cleared by default.

Also, to save some work for the Tier2 SDK, we should rename this method to flushDistributions() and should return a list of distribution strings - List<Pair<Long, String> centroids
where the 2 tuple Pair represents - distribution

that way the Tier 2 client does not have to do this and need not know what MinuteBin is -

for (MinuteBin minuteBin : bins) {
          List<Pair<Double, Integer>> centroids = new ArrayList<>();
          for (Centroid c : minuteBin.getDist().centroids()) {
            centroids.add(new Pair<>(c.mean(), c.count()));
          }
          wavefront.send(histogramGranularities, minuteBin.getMinuteMillis() / 1000,
              centroids, prefixAndSanitize(name), source, pointTags);
        }

Also mention in the javadoc that flushDistributions() has side effects (i.e. it is not read-only method) and will mutate the state of your system.